rc/filetype/kotlin.kak

```diff
@@ -0,0 +1,299 @@
+add-highlighter shared/ktor regions
+add-highlighter shared/ktor/code default-region group
+add-highlighter shared/ktor/comment      region /\* \*/ fill comment
+add-highlighter shared/ktor/line_comment region // $ fill comment
+
+add-highlighter shared/ktor/code/ktorClient regex \b(HttpClient)\b 1:red+u
+add-highlighter shared/ktor/code/ktorCall   regex \b(Android|Apache|CIO|Ios|Jetty|Netty|OkHttp|Tomcat)\b 1:green+u
+add-highlighter shared/ktor/code/ktorType   regex \b(ApplicationCall|ApplicationSendPipeline|ByteReadChannel|Headers|HeadersBuilder|HttpClientCall|HttpClientEngine|HttpMessageBuilder|HttpMethod|HttpProtocolVersion|HttpRequestBuilder|HttpResponse|HttpStatusCode|ResponseCookies|ResponseHeaders|URLBuilder)\b 1:yellow+u
+
+
+# KDOC <https://kotlinlang.org/docs/reference/kotlin-doc.html>
+add-highlighter shared/kdoc group
+add-highlighter shared/kdoc/tag regex \*(?:\s+)?(@(author|constructor|exception|param|property|receiver|return|sample|see|since|suppress|throws))\b 1:default+ui
+
+
+# DISCOLOUR
+add-highlighter shared/kotlin/code/discolour regex ^(package|import)(?S)(.+) 2:default+fa
+
+
+# Commands
+# ‾‾‾‾‾‾‾‾
+
+define-command -hidden kotlin-indent-on-new-line %~
+  evaluate-commands -draft -itersel %<
+    # preserve previous line indent
+    try %{ execute-keys -draft <semicolon>K<a-&> }
+    # indent after lines ending with { or (
+    try %[ execute-keys -draft k<a-x> <a-k> [{(]\h*$ <ret> j<a-gt> ]
+    # cleanup trailing white spaces on the previous line
+    try %{ execute-keys -draft k<a-x> s \h+$ <ret>d }
+    # align to opening paren of previous line
+    try %{ execute-keys -draft [( <a-k> \A\([^\n]+\n[^\n]*\n?\z <ret> s \A\(\h*.|.\z <ret> '<a-;>' & }
+    # copy // comments prefix
+    try %{ execute-keys -draft <semicolon><c-s>k<a-x> s ^\h*\K/{2,} <ret> y<c-o>P<esc> }
+    # indent after a pattern match on when/where statements
+    try %[ execute-keys -draft k<a-x> <a-k> ^\h*(when|where).*$ <ret> j<a-gt> ]
+    # indent after term on an expression
+    try %[ execute-keys -draft k<a-x> <a-k> =\h*?$ <ret> j<a-gt> ]
+    # indent after keywords
+    try %[ execute-keys -draft <semicolon><a-F>)MB <a-k> \A(catch|do|else|for|if|try|while)\h*\(.*\)\h*\n\h*\n?\z <ret> s \A|.\z <ret> 1<a-&>1<a-space><a-gt> ]
+    # deindent closing brace(s) when after cursor
+    try %[ execute-keys -draft <a-x> <a-k> ^\h*[})] <ret> gh / [})] <ret> m <a-S> 1<a-&> ]
+  >
+~
+
+define-command -hidden kotlin-indent-on-opening-curly-brace %[
+  # align indent with opening paren when { is entered on a new line after the closing paren
+  try %[ execute-keys -draft -itersel h<a-F>)M <a-k> \A\(.*\)\h*\n\h*\{\z <ret> s \A|.\z <ret> 1<a-&> ]
+]
+
+define-command -hidden kotlin-indent-on-closing-curly-brace %[
+  # align to opening curly brace when alone on a line
+  try %[ execute-keys -itersel -draft <a-h><a-k>^\h+\}$<ret>hms\A|.\z<ret>1<a-&> ]
+]
+
+
+# Exceptions, Errors, and Types
+# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
+
+evaluate-commands %sh{
```

Comment:
That seems wildly over-engineered, why not stick to pure shell scripting?

Reply:
Do you mean usage of awk in general? or the script itself?
As far as I understand, it's trying to split a long space separated string into a list which then gets consumed in the highlighter function.

Reply:
This particular use of Awk seems overkill, it doesn't look like it's more readable, faster, shorter than how other scripts handle that.

For example, `go.kak`:

https://github.com/mawww/kakoune/blob/7c321e121d/rc/filetype/go.kak#L54-L78